In [29]:
import pandas as pd
import numpy as np


$$cov(X, Y) = \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}{n-1}$$

## Implemention Covariance (No NumPy)

In [30]:
a = [4, 5, 12, 7]
b = [1, 5, 3, 9]

In [61]:
n = len(a) - 1

In [31]:
def mean(xs):
    return sum(xs) / len(xs) 

In [32]:
mean_a = mean(a)
mean_b = mean(b)
print(mean_b)

4.5


In [33]:
def plus_array(mean, xs):
    print('mean ->', mean)
    print('array ->',xs)
    centered = []
    
    for x in xs:
        centered.append((x - mean))
        
    return centered

In [34]:
centered_a = plus_array(mean_a, a)
centered_b = plus_array(mean_b, b)

mean -> 7.0
array -> [4, 5, 12, 7]
mean -> 4.5
array -> [1, 5, 3, 9]


In [35]:
def print_centered(og, centered):
    print(f'array original: {og} -> after centered: {centered}' )

In [47]:
print_centered(a,centered_a)
print_centered(b, centered_b)

array original: [4, 5, 12, 7] -> after centered: [-3.0, -2.0, 5.0, 0.0]
array original: [1, 5, 3, 9] -> after centered: [-3.5, 0.5, -1.5, 4.5]


In [54]:
def multiply_chain(x, y):
    if len(x) != len(y):
        print('Erro: Input vectors must have the same dimensions')
        return
   
    length = len(x)
   
    sum_of_products = []
   
    print(f'{x} -> {y}')
   
    for i in range(0, length):
       print(f'{x[i]} * {y[i]} = {(x[i] * y[i])}')
       sum_of_products.append((x[i] * y[i]))
       
    
    print(f"sum_of_products: {sum_of_products}")
    return sum_of_products

In [56]:
def sum_products(xi):
    aux = 0
    for x in xi:
        aux += x
    return aux

In [57]:
sum_of_products = sum_products(multiply_chain(centered_a, centered_b))

[-3.0, -2.0, 5.0, 0.0] -> [-3.5, 0.5, -1.5, 4.5]
-3.0 * -3.5 = 10.5
-2.0 * 0.5 = -1.0
5.0 * -1.5 = -7.5
0.0 * 4.5 = 0.0
sum_of_products: [10.5, -1.0, -7.5, 0.0]


In [58]:
print(sum_of_products)

2.0


In [62]:

covariance = sum_of_products / n

In [63]:
print(covariance)

0.6666666666666666


### Covariance Calculation
**Formula:**
$$cov(X, Y) = \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}{n-1}$$

**Steps for $a=[4, 5, 12, 7]$ and $b=[1, 5, 3, 9]$:**
1. Means: $\bar{a}=7.0$, $\bar{b}=4.5$
2. Sum of cross-products: $(10.5) + (-1.0) + (-7.5) + (0) = 2.0$
3. Division ($n-1$): $2.0 / 3 \approx 0.6667$